File: src/features/admin/services/adminService.ts
import { mockAdapter } from '@/lib/api/mockAdapter'
import { apiClient } from '@/lib/api/apiClient'
import { endpoints } from '@/lib/api/endpoints'
import { shouldUseApi } from '@/lib/api/config'
import { STORAGE_KEYS } from '@/lib/constants'
import { readStorage, writeStorage } from '@/lib/storage/localStorage'
import { uid } from '@/lib/utils'
import type {
  AdminActivity,
  AdminOverview,
  AdminUserCreatePayload,
  AdminUserItem,
  AdminUserUpdatePayload,
  SystemStatus,
} from '@/types/admin'
import type {
  DependenciaMateria,
  HistorialAcademico,
  InscripcionPrograma,
  Materia,
  Programa,
  VersionMalla,
} from '@/types/curriculum'
import type { Estudiante, Usuario } from '@/types/auth'
import { estudiantesMock, usuariosMock } from '@/mocks/users.mock'

const activities: AdminActivity[] = [
  {
    id: 'act_1',
    descripcion: 'Nueva materia creada: Seguridad Informática',
    fecha: '2026-05-14T15:00:00.000Z',
    tipo: 'materia',
  },
  {
    id: 'act_2',
    descripcion: 'Malla 2025 de Ingeniería de Sistemas actualizada',
    fecha: '2026-05-13T12:00:00.000Z',
    tipo: 'malla',
  },
  {
    id: 'act_3',
    descripcion: 'Prerrequisito modificado: INF202 requiere INF201',
    fecha: '2026-05-12T09:30:00.000Z',
    tipo: 'dependencia',
  },
  {
    id: 'act_4',
    descripcion: 'Estudiante transferido a doble programa',
    fecha: '2026-05-11T10:45:00.000Z',
    tipo: 'estudiante',
  },
]

export const adminService = {
  getOverview() {
    if (shouldUseApi()) {
      return apiClient.get<AdminOverview>(endpoints.admin.overview)
    }
    return mockAdapter<AdminOverview>(() => ({
      estudiantesActivos: 2,
      programas: readStorage<Programa[]>(STORAGE_KEYS.programs, []).length,
      totalMaterias: readStorage<Materia[]>(STORAGE_KEYS.courses, []).length,
      completitudPromedio: 82,
    }))
  },
  getDashboardData() {
    if (shouldUseApi()) {
      return apiClient.get<{
        programas: Programa[]
        materias: Materia[]
        versiones: VersionMalla[]
        dependencias: DependenciaMateria[]
        activities: AdminActivity[]
      }>(endpoints.admin.dashboard)
    }
    return mockAdapter(() => ({
      programas: readStorage<Programa[]>(STORAGE_KEYS.programs, []),
      materias: readStorage<Materia[]>(STORAGE_KEYS.courses, []),
      versiones: readStorage<VersionMalla[]>(STORAGE_KEYS.versions, []),
      dependencias: readStorage<DependenciaMateria[]>(
        STORAGE_KEYS.dependencies,
        [],
      ),
      activities,
    }))
  },
  getSystemStatus() {
    if (shouldUseApi()) {
      return apiClient.get<SystemStatus>(endpoints.admin.systemStatus)
    }
    return mockAdapter<SystemStatus>(() => {
      const settings = readStorage<Record<string, unknown>>(STORAGE_KEYS.settings, {})
      return {
        environment: 'mock',
        appName: 'CurriculaPath API simulada',
        checks: [
          {
            id: 'api',
            nombre: 'API REST',
            estado: 'ok',
            detalle: 'Modo mock activo: servicios desacoplados responden desde localStorage.',
          },
          {
            id: 'database',
            nombre: 'Base de datos',
            estado: 'warning',
            detalle: 'Persistencia temporal en localStorage; lista para sustituirse por API real.',
            accionRecomendada: 'Usar fuente de datos API real para persistencia multiusuario.',
          },
          {
            id: 'ocr',
            nombre: 'OCR local',
            estado: 'warning',
            detalle: 'Pipeline visual disponible; OCR real se valida desde el backend.',
            accionRecomendada: 'Activar API real y revisar diagnóstico OCR.',
          },
          {
            id: 'llm',
            nombre: 'LLM local / RAG',
            estado: settings.llmConnected ? 'ok' : 'warning',
            detalle: settings.llmConnected
              ? 'Modelo local marcado como conectado en configuración.'
              : 'Chat/RAG opera con respuestas mock fundamentadas.',
            accionRecomendada: settings.llmConnected
              ? null
              : 'Probar conexión con Gemma, Llama o Mistral local.',
          },
          {
            id: 'security',
            nombre: 'Seguridad HTTP',
            estado: 'warning',
            detalle: 'Headers y HTTPS se aplican en la API real, no en modo mock.',
            accionRecomendada: 'Validar docs/SECURITY_AND_OPERATIONS.md para despliegue.',
          },
          {
            id: 'configuration',
            nombre: 'Configuración',
            estado: 'ok',
            detalle: 'Preferencias locales cargadas correctamente.',
          },
        ],
      }
    })
  },
  getUsers() {
    if (shouldUseApi()) {
      return apiClient.get<AdminUserItem[]>(endpoints.admin.users)
    }
    return mockAdapter<AdminUserItem[]>(() => {
      const users = readStorage<Usuario[]>(STORAGE_KEYS.users, usuariosMock)
      const students = readStorage<Estudiante[]>(STORAGE_KEYS.students, estudiantesMock)
      return users.map((user) => ({
        user,
        student: students.find((student) => student.usuarioId === user.id) ?? null,
      }))
    })
  },
  createUser(payload: AdminUserCreatePayload) {
    if (shouldUseApi()) {
      return apiClient.post<AdminUserCreatePayload, AdminUserItem>(
        endpoints.admin.users,
        payload,
      )
    }
    return mockAdapter<AdminUserItem>(() => {
      const users = readStorage<Usuario[]>(STORAGE_KEYS.users, usuariosMock)
      if (users.some((user) => user.email.toLowerCase() === payload.email.toLowerCase())) {
        throw new Error('Ya existe un usuario con ese email')
      }
      const now = new Date().toISOString()
      const user: Usuario = {
        id: uid('user'),
        nombre: payload.nombre,
        email: payload.email,
        rol: payload.rol,
        activo: payload.activo,
        creadoEn: now,
      }
      let student: Estudiante | null = null
      writeStorage(STORAGE_KEYS.users, [...users, user])
      writeStorage(STORAGE_KEYS.mockPasswords, {
        ...readStorage<Record<string, string>>(STORAGE_KEYS.mockPasswords, {}),
        [payload.email.toLowerCase()]: payload.password,
      })

      if (payload.rol === 'student') {
        if (!payload.codigoEstudiantil || !payload.programaPrincipalId) {
          throw new Error('Los estudiantes requieren código y programa principal')
        }
        const students = readStorage<Estudiante[]>(STORAGE_KEYS.students, estudiantesMock)
        if (students.some((item) => item.codigoEstudiantil === payload.codigoEstudiantil)) {
          throw new Error('Ya existe un estudiante con ese código')
        }
        student = {
          id: uid('student'),
          usuarioId: user.id,
          codigoEstudiantil: payload.codigoEstudiantil,
          semestreActual: payload.semestreActual ?? 1,
          cargaMaximaCreditos: payload.cargaMaximaCreditos ?? 20,
          creadoEn: now,
        }
        writeStorage(STORAGE_KEYS.students, [...students, student])
        const selectedPrograms = [
          payload.programaPrincipalId,
          ...(payload.programaSecundarioId ? [payload.programaSecundarioId] : []),
        ]
        const enrollments = readStorage<InscripcionPrograma[]>(STORAGE_KEYS.enrollments, [])
        writeStorage(STORAGE_KEYS.enrollments, [
          ...enrollments,
          ...selectedPrograms.map((programId) => ({
            id: uid('enroll'),
            estudianteId: student!.id,
            programaId: programId,
            esPrincipal: programId === payload.programaPrincipalId,
            fechaInscripcion: now,
          })),
        ])
        const versions = readStorage<VersionMalla[]>(STORAGE_KEYS.versions, [])
        const courses = readStorage<Materia[]>(STORAGE_KEYS.courses, [])
        const histories = readStorage<HistorialAcademico[]>(STORAGE_KEYS.histories, [])
        const newHistories = selectedPrograms.flatMap((programId) => {
          const version = versions.find((item) => item.programaId === programId && item.activa)
          return courses
            .filter((course) => course.versionMallaId === version?.id)
            .map<HistorialAcademico>((course) => ({
              id: uid('history'),
              estudianteId: student!.id,
              materiaId: course.id,
              estado: 'pendiente',
              actualizadoEn: now,
            }))
        })
        writeStorage(STORAGE_KEYS.histories, [...histories, ...newHistories])
      }

      return { user, student }
    })
  },
  updateUser(userId: string, payload: AdminUserUpdatePayload) {
    if (shouldUseApi()) {
      return apiClient.patch<AdminUserUpdatePayload, AdminUserItem>(
        `${endpoints.admin.users}/${userId}`,
        payload,
      )
    }
    return mockAdapter<AdminUserItem>(() => {
      const users = readStorage<Usuario[]>(STORAGE_KEYS.users, usuariosMock)
      const next = users.map((user) => (user.id === userId ? { ...user, ...payload } : user))
      writeStorage(STORAGE_KEYS.users, next)
      const user = next.find((item) => item.id === userId)
      if (!user) throw new Error('Usuario no encontrado')
      const student =
        readStorage<Estudiante[]>(STORAGE_KEYS.students, estudiantesMock).find(
          (item) => item.usuarioId === user.id,
        ) ?? null
      return { user, student }
    })
  },
  resetUserPassword(userId: string, newPassword: string) {
    if (shouldUseApi()) {
      return apiClient.post<{ newPassword: string }, { ok: boolean }>(
        endpoints.admin.resetUserPassword(userId),
        { newPassword },
      )
    }
    return mockAdapter(() => {
      const user = readStorage<Usuario[]>(STORAGE_KEYS.users, usuariosMock).find(
        (item) => item.id === userId,
      )
      if (!user) throw new Error('Usuario no encontrado')
      writeStorage(STORAGE_KEYS.mockPasswords, {
        ...readStorage<Record<string, string>>(STORAGE_KEYS.mockPasswords, {}),
        [user.email.toLowerCase()]: newPassword,
      })
      return { ok: true }
    })
  },
  createProgram(payload: Omit<Programa, 'id'>) {
    if (shouldUseApi()) {
      return apiClient.post<Omit<Programa, 'id'>, Programa>(
        endpoints.programs,
        payload,
      )
    }
    return mockAdapter(() => {
      const programs = readStorage<Programa[]>(STORAGE_KEYS.programs, [])
      const program: Programa = { ...payload, id: uid('program') }
      writeStorage(STORAGE_KEYS.programs, [...programs, program])
      return program
    })
  },
  updateProgram(programId: string, payload: Partial<Omit<Programa, 'id'>>) {
    if (shouldUseApi()) {
      return apiClient.patch<Partial<Omit<Programa, 'id'>>, Programa>(
        `${endpoints.programs}/${programId}`,
        payload,
      )
    }
    return mockAdapter(() => {
      const programs = readStorage<Programa[]>(STORAGE_KEYS.programs, [])
      const next = programs.map((program) =>
        program.id === programId ? { ...program, ...payload } : program,
      )
      writeStorage(STORAGE_KEYS.programs, next)
      return next.find((item) => item.id === programId)
    })
  },
  toggleProgram(programId: string) {
    if (shouldUseApi()) {
      return adminService.getDashboardData().then((dashboard) => {
        const program = dashboard.programas.find((item) => item.id === programId)
        if (!program) throw new Error('Programa no encontrado')
        return adminService.updateProgram(programId, { activo: !program.activo })
      })
    }
    return mockAdapter(() => {
      const programs = readStorage<Programa[]>(STORAGE_KEYS.programs, [])
      const next = programs.map((program) =>
        program.id === programId ? { ...program, activo: !program.activo } : program,
      )
      writeStorage(STORAGE_KEYS.programs, next)
      return next.find((item) => item.id === programId)
    })
  },
  createCourse(payload: Omit<Materia, 'id'>) {
    if (shouldUseApi()) {
      return apiClient.post<Omit<Materia, 'id'>, Materia>(
        endpoints.courses,
        payload,
      )
    }
    return mockAdapter(() => {
      const courses = readStorage<Materia[]>(STORAGE_KEYS.courses, [])
      const course: Materia = { ...payload, id: uid('course') }
      writeStorage(STORAGE_KEYS.courses, [...courses, course])
      return course
    })
  },
  updateCourse(courseId: string, payload: Partial<Omit<Materia, 'id'>>) {
    if (shouldUseApi()) {
      return apiClient.patch<Partial<Omit<Materia, 'id'>>, Materia>(
        `${endpoints.courses}/${courseId}`,
        payload,
      )
    }
    return mockAdapter(() => {
      const courses = readStorage<Materia[]>(STORAGE_KEYS.courses, [])
      const next = courses.map((course) =>
        course.id === courseId ? { ...course, ...payload } : course,
      )
      writeStorage(STORAGE_KEYS.courses, next)
      return next.find((item) => item.id === courseId)
    })
  },
  createVersion(payload: Omit<VersionMalla, 'id'>) {
    if (shouldUseApi()) {
      return apiClient.post<Omit<VersionMalla, 'id'>, VersionMalla>(
        endpoints.admin.versions,
        payload,
      )
    }
    return mockAdapter(() => {
      const versions = readStorage<VersionMalla[]>(STORAGE_KEYS.versions, [])
      const version: VersionMalla = { ...payload, id: uid('version') }
      const next = payload.activa
        ? [
            ...versions.map((item) =>
              item.programaId === payload.programaId
                ? { ...item, activa: false }
                : item,
            ),
            version,
          ]
        : [...versions, version]
      writeStorage(STORAGE_KEYS.versions, next)
      return version
    })
  },
  updateVersion(
    versionId: string,
    payload: Partial<Omit<VersionMalla, 'id'>>,
  ) {
    if (shouldUseApi()) {
      return apiClient.patch<
        Partial<Omit<VersionMalla, 'id'>>,
        VersionMalla
      >(`${endpoints.admin.versions}/${versionId}`, payload)
    }
    return mockAdapter(() => {
      const versions = readStorage<VersionMalla[]>(STORAGE_KEYS.versions, [])
      const current = versions.find((item) => item.id === versionId)
      if (!current) throw new Error('Versión no encontrada')
      const next = versions.map((version) =>
        version.id === versionId ? { ...version, ...payload } : version,
      )
      writeStorage(STORAGE_KEYS.versions, next)
      return next.find((item) => item.id === versionId)
    })
  },
  toggleVersion(versionId: string) {
    if (shouldUseApi()) {
      return apiClient.post<Record<string, never>, VersionMalla>(
        `${endpoints.admin.versions}/${versionId}/toggle`,
        {},
      )
    }
    return mockAdapter(() => {
      const versions = readStorage<VersionMalla[]>(STORAGE_KEYS.versions, [])
      const target = versions.find((version) => version.id === versionId)
      if (!target) throw new Error('Versión no encontrada')
      const activeSiblings = versions.filter(
        (version) =>
          version.programaId === target.programaId && version.activa,
      )
      if (target.activa && activeSiblings.length <= 1) {
        throw new Error('Debe existir al menos una versión activa por programa')
      }
      const next = versions.map((version) => {
        if (version.programaId !== target.programaId) return version
        if (version.id === versionId) {
          return { ...version, activa: !version.activa }
        }
        return target.activa ? version : { ...version, activa: false }
      })
      writeStorage(STORAGE_KEYS.versions, next)
      return next.find((item) => item.id === versionId)
    })
  },
  createDependency(payload: Omit<DependenciaMateria, 'id'>) {
    if (shouldUseApi()) {
      return apiClient.post<Omit<DependenciaMateria, 'id'>, DependenciaMateria>(
        endpoints.admin.dependencies,
        payload,
      )
    }
    return mockAdapter(() => {
      const dependencies = readStorage<DependenciaMateria[]>(
        STORAGE_KEYS.dependencies,
        [],
      )
      const duplicated = dependencies.some(
        (dependency) =>
          dependency.materiaId === payload.materiaId &&
          dependency.materiaRequeridaId === payload.materiaRequeridaId &&
          dependency.tipo === payload.tipo,
      )
      if (duplicated) {
        throw new Error('La dependencia ya existe')
      }
      const dependency: DependenciaMateria = { ...payload, id: uid('dependency') }
      writeStorage(STORAGE_KEYS.dependencies, [...dependencies, dependency])
      return dependency
    })
  },
  deleteDependency(dependencyId: string) {
    if (shouldUseApi()) {
      return apiClient.delete<{ dependencyId: string; deleted: boolean }>(
        `${endpoints.admin.dependencies}/${dependencyId}`,
      )
    }
    return mockAdapter(() => {
      const dependencies = readStorage<DependenciaMateria[]>(
        STORAGE_KEYS.dependencies,
        [],
      )
      const next = dependencies.filter((dependency) => dependency.id !== dependencyId)
      writeStorage(STORAGE_KEYS.dependencies, next)
      return { dependencyId, deleted: true }
    })
  },
}

File: backend/app/services/admin_service.py
from uuid import uuid4

from fastapi import HTTPException, status
from sqlalchemy import func, select, text
from sqlalchemy.orm import Session

from app.core.config import get_settings
from app.models.entities import (
    AdminActivity,
    Course,
    CourseDependency,
    CurriculumVersion,
    Program,
    Student,
)
from app.models.enums import DependencyType
from app.schemas.admin import (
    AdminDashboardResponse,
    AdminOverviewResponse,
    SystemCheckResponse,
    SystemStatusResponse,
    VersionCreate,
    VersionUpdate,
)
from app.schemas.curriculum import CourseCreate, CourseUpdate, ProgramCreate, ProgramUpdate
from app.services.pdf_processing import get_ocr_status


def _record_activity(db: Session, descripcion: str, tipo: str) -> None:
    db.add(
        AdminActivity(
            id=f"activity_{uuid4().hex[:10]}",
            descripcion=descripcion,
            tipo=tipo,
        ),
    )


def _ensure_program_code_available(
    db: Session,
    codigo: str,
    *,
    exclude_program_id: str | None = None,
) -> None:
    query = select(Program).where(func.lower(Program.codigo) == codigo.lower())
    if exclude_program_id:
        query = query.where(Program.id != exclude_program_id)
    if db.scalar(query):
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
            detail="Ya existe un programa con ese código",
        )


def _ensure_course_code_available(
    db: Session,
    version_malla_id: str,
    codigo: str,
    *,
    exclude_course_id: str | None = None,
) -> None:
    query = select(Course).where(
        Course.version_malla_id == version_malla_id,
        func.lower(Course.codigo) == codigo.lower(),
    )
    if exclude_course_id:
        query = query.where(Course.id != exclude_course_id)
    if db.scalar(query):
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
            detail="Ya existe una materia con ese código en la versión seleccionada",
        )


def _get_version_or_404(db: Session, version_id: str) -> CurriculumVersion:
    version = db.get(CurriculumVersion, version_id)
    if not version:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Versión no encontrada")
    return version


def _get_program_or_404(db: Session, program_id: str) -> Program:
    program = db.get(Program, program_id)
    if not program:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Programa no encontrado")
    return program


def _get_course_or_404(db: Session, course_id: str) -> Course:
    course = db.get(Course, course_id)
    if not course:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Materia no encontrada")
    return course


def _creates_prerequisite_cycle(
    db: Session,
    materia_id: str,
    materia_requerida_id: str,
) -> bool:
    prerequisites = list(
        db.scalars(
            select(CourseDependency).where(CourseDependency.tipo == DependencyType.PREREQUISITO),
        ),
    )
    adjacency: dict[str, list[str]] = {}
    for dependency in prerequisites:
        adjacency.setdefault(dependency.materia_id, []).append(dependency.materia_requerida_id)

    pending = [materia_requerida_id]
    visited: set[str] = set()
    while pending:
        current = pending.pop()
        if current == materia_id:
            return True
        if current in visited:
            continue
        visited.add(current)
        pending.extend(adjacency.get(current, []))
    return False


def get_overview(db: Session) -> AdminOverviewResponse:
    students = db.scalar(select(func.count()).select_from(Student)) or 0
    programs = db.scalar(select(func.count()).select_from(Program)) or 0
    courses = db.scalar(select(func.count()).select_from(Course)) or 0
    return AdminOverviewResponse(
        estudiantes_activos=students,
        programas=programs,
        total_materias=courses,
        completitud_promedio=82,
    )


def get_dashboard_data(db: Session) -> AdminDashboardResponse:
    return AdminDashboardResponse(
        programas=list(db.scalars(select(Program).order_by(Program.nombre))),
        materias=list(db.scalars(select(Course).order_by(Course.codigo))),
        versiones=list(db.scalars(select(CurriculumVersion).order_by(CurriculumVersion.id))),
        dependencias=list(db.scalars(select(CourseDependency).order_by(CourseDependency.id))),
        activities=list(
            db.scalars(
                select(AdminActivity).order_by(AdminActivity.fecha.desc()).limit(8),
            ),
        ),
    )


def get_system_status(db: Session) -> SystemStatusResponse:
    settings = get_settings()
    checks: list[SystemCheckResponse] = []

    checks.append(
        SystemCheckResponse(
            id="api",
            nombre="API REST",
            estado="ok",
            detalle=f"{settings.app_name} responde en entorno {settings.environment}.",
        ),
    )

    try:
        db.execute(text("SELECT 1"))
        checks.append(
            SystemCheckResponse(
                id="database",
                nombre="Base de datos",
                estado="ok",
                detalle="La conexión de base de datos respondió correctamente.",
            ),
        )
    except Exception as exc:
        checks.append(
            SystemCheckResponse(
                id="database",
                nombre="Base de datos",
                estado="error",
                detalle=f"No se pudo validar la base de datos: {exc}",
                accion_recomendada="Revisar DATABASE_URL, migraciones y disponibilidad del motor.",
            ),
        )

    ocr = get_ocr_status()
    checks.append(
        SystemCheckResponse(
            id="ocr",
            nombre="OCR local",
            estado="ok" if ocr.ready_for_scanned_pdfs else ("warning" if ocr.available else "error"),
            detalle=ocr.message,
            accion_recomendada=" ".join(ocr.next_steps) if ocr.next_steps else None,
        ),
    )

    checks.append(
        SystemCheckResponse(
            id="llm",
            nombre="LLM local / RAG",
            estado="warning",
            detalle=(
                "RAG está operativo con contexto académico. El LLM local se valida desde "
                "Configuración o desde /llm/connect para no bloquear el panel administrativo."
            ),
            accion_recomendada=(
                "Configurar Gemma, Llama o Mistral local si se requiere generación real "
                "sin fallback mock."
            ),
        ),
    )

    security_ok = settings.secure_headers_enabled and (
        not settings.is_production or settings.enforce_https
    )
    checks.append(
        SystemCheckResponse(
            id="security",
            nombre="Seguridad HTTP",
            estado="ok" if security_ok else "warning",
            detalle=(
                "Headers de seguridad activos"
                if settings.secure_headers_enabled
                else "Headers de seguridad desactivados"
            )
            + (
                " y HTTPS exigido."
                if settings.enforce_https
                else "; HTTPS no forzado en este entorno."
            ),
            accion_recomendada=(
                None
                if security_ok
                else "Activar SECURE_HEADERS_ENABLED y ENFORCE_HTTPS antes de producción."
            ),
        ),
    )

    try:
        settings.validate_runtime_security()
        config_state = "ok"
        config_detail = "La configuración actual cumple las validaciones del entorno."
        config_action = None
    except RuntimeError as exc:
        config_state = "error"
        config_detail = str(exc)
        config_action = "Corregir variables de entorno antes de publicar."

    checks.append(
        SystemCheckResponse(
            id="configuration",
            nombre="Configuración",
            estado=config_state,
            detalle=config_detail,
            accion_recomendada=config_action,
        ),
    )

    return SystemStatusResponse(
        environment=settings.environment,
        app_name=settings.app_name,
        checks=checks,
    )


def create_program(db: Session, payload: ProgramCreate) -> Program:
    _ensure_program_code_available(db, payload.codigo)
    program = Program(id=f"prog_{uuid4().hex[:10]}", **payload.model_dump())
    db.add(program)
    _record_activity(db, f"Programa creado: {program.codigo} · {program.nombre}", "malla")
    db.commit()
    db.refresh(program)
    return program


def update_program(db: Session, program_id: str, payload: ProgramUpdate) -> Program:
    program = _get_program_or_404(db, program_id)
    updates = payload.model_dump(exclude_unset=True)
    if "codigo" in updates and updates["codigo"] is not None:
        _ensure_program_code_available(db, updates["codigo"], exclude_program_id=program.id)
    for key, value in updates.items():
        setattr(program, key, value)
    db.add(program)
    _record_activity(db, f"Programa actualizado: {program.codigo} · {program.nombre}", "malla")
    db.commit()
    db.refresh(program)
    return program


def create_course(db: Session, payload: CourseCreate) -> Course:
    _get_version_or_404(db, payload.version_malla_id)
    _ensure_course_code_available(db, payload.version_malla_id, payload.codigo)
    course = Course(id=f"course_{uuid4().hex[:10]}", **payload.model_dump())
    db.add(course)
    _record_activity(db, f"Materia creada: {course.codigo} · {course.nombre}", "materia")
    db.commit()
    db.refresh(course)
    return course


def update_course(db: Session, course_id: str, payload: CourseUpdate) -> Course:
    course = _get_course_or_404(db, course_id)
    updates = payload.model_dump(exclude_unset=True)
    next_version_id = updates.get("version_malla_id", course.version_malla_id)
    next_codigo = updates.get("codigo", course.codigo)
    _get_version_or_404(db, next_version_id)
    _ensure_course_code_available(
        db,
        next_version_id,
        next_codigo,
        exclude_course_id=course.id,
    )
    for key, value in updates.items():
        setattr(course, key, value)
    db.add(course)
    _record_activity(db, f"Materia actualizada: {course.codigo} · {course.nombre}", "materia")
    db.commit()
    db.refresh(course)
    return course


def create_version(db: Session, payload: VersionCreate) -> CurriculumVersion:
    _get_program_or_404(db, payload.programa_id)
    if payload.activa:
        current_versions = list(
            db.scalars(
                select(CurriculumVersion).where(CurriculumVersion.programa_id == payload.programa_id),
            ),
        )
        for version in current_versions:
            version.activa = False
            db.add(version)
    version = CurriculumVersion(id=f"version_{uuid4().hex[:10]}", **payload.model_dump())
    db.add(version)
    _record_activity(db, f"Versión de malla creada: {version.nombre_version}", "malla")
    db.commit()
    db.refresh(version)
    return version


def update_version(db: Session, version_id: str, payload: VersionUpdate) -> CurriculumVersion:
    version = _get_version_or_404(db, version_id)
    updates = payload.model_dump(exclude_unset=True)
    target_program_id = updates.get("programa_id", version.programa_id)
    _get_program_or_404(db, target_program_id)

    if (
        target_program_id != version.programa_id
        and version.activa
        and not db.scalar(
            select(CurriculumVersion).where(
                CurriculumVersion.programa_id == version.programa_id,
                CurriculumVersion.id != version.id,
                CurriculumVersion.activa.is_(True),
            ),
        )
    ):
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
            detail="No puedes mover la única versión activa de un programa",
        )

    should_be_active = updates.get("activa", version.activa)
    if should_be_active:
        siblings = list(
            db.scalars(
                select(CurriculumVersion).where(
                    CurriculumVersion.programa_id == target_program_id,
                    CurriculumVersion.id != version.id,
                ),
            ),
        )
        for sibling in siblings:
            sibling.activa = False
            db.add(sibling)
    for key, value in updates.items():
        setattr(version, key, value)
    db.add(version)
    _record_activity(db, f"Versión de malla actualizada: {version.nombre_version}", "malla")
    db.commit()
    db.refresh(version)
    return version


def toggle_version(db: Session, version_id: str) -> CurriculumVersion:
    version = _get_version_or_404(db, version_id)
    siblings = list(
        db.scalars(
            select(CurriculumVersion).where(CurriculumVersion.programa_id == version.programa_id),
        ),
    )
    active_siblings = [item for item in siblings if item.activa]
    if version.activa and len(active_siblings) <= 1:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
            detail="Debe existir al menos una versión activa por programa",
        )
    version.activa = not version.activa
    if version.activa:
        for sibling in siblings:
            if sibling.id != version.id:
                sibling.activa = False
                db.add(sibling)
    db.add(version)
    _record_activity(
        db,
        f"Versión de malla {'activada' if version.activa else 'desactivada'}: {version.nombre_version}",
        "malla",
    )
    db.commit()
    db.refresh(version)
    return version


def create_dependency(
    db: Session,
    materia_id: str,
    materia_requerida_id: str,
    tipo: DependencyType,
) -> CourseDependency:
    materia = _get_course_or_404(db, materia_id)
    materia_requerida = _get_course_or_404(db, materia_requerida_id)
    if materia.id == materia_requerida.id:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
            detail="Una materia no puede depender de sí misma",
        )
    if materia.version_malla_id != materia_requerida.version_malla_id:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
            detail="Las dependencias solo pueden crearse dentro de la misma versión de malla",
        )
    duplicated = db.scalar(
        select(CourseDependency).where(
            CourseDependency.materia_id == materia_id,
            CourseDependency.materia_requerida_id == materia_requerida_id,
            CourseDependency.tipo == tipo,
        ),
    )
    if duplicated:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
            detail="La dependencia ya existe",
        )
    if tipo == DependencyType.PREREQUISITO and _creates_prerequisite_cycle(
        db,
        materia_id,
        materia_requerida_id,
    ):
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
            detail="La dependencia crearía un ciclo de prerrequisitos",
        )
    dependency = CourseDependency(
        id=f"dependency_{uuid4().hex[:10]}",
        materia_id=materia_id,
        materia_requerida_id=materia_requerida_id,
        tipo=tipo,
    )
    db.add(dependency)
    _record_activity(
        db,
        f"Dependencia creada: {materia.codigo} requiere {materia_requerida.codigo}",
        "dependencia",
    )
    db.commit()
    db.refresh(dependency)
    return dependency


def delete_dependency(db: Session, dependency_id: str) -> dict[str, str | bool]:
    dependency = db.get(CourseDependency, dependency_id)
    if not dependency:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Dependencia no encontrada")
    materia = _get_course_or_404(db, dependency.materia_id)
    materia_requerida = _get_course_or_404(db, dependency.materia_requerida_id)
    db.delete(dependency)
    _record_activity(
        db,
        f"Dependencia eliminada: {materia.codigo} ya no requiere {materia_requerida.codigo}",
        "dependencia",
    )
    db.commit()
    return {"dependencyId": dependency_id, "deleted": True}
